### PR TITLE
Mobile phone services: SigninOrRegisterResidentService, ChangePhoneNumberResidentService

### PR DESCRIPTION
**Pull request by dkoviazin**
The User now will have a type 'resident' or 'staff'

### CODE REVIEW

File: apps/condo/domains/user/access/ChangePhoneNumberResidentService.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
+/**
+ * Generated by `createservice user.ChangePhoneNumberResidentService`
+ */
+const { throwAuthenticationError } = require('@condo/domains/common/utils/apolloErrorFormatter')
+
+async function canChangePhoneNumberResident ({ authentication: { item: user } }) {
+    if (!user) return throwAuthenticationError()
+    console.log('user', user)
```
Comment (SavelevMatthew) [inline apps/condo/domains/user/access/ChangePhoneNumberResidentService.js:8]: Debugging console log? Don't forget to remove it later :)

**Review by AntonAL — COMMENTED** [review]
My suggestions to prevent naming clash between `User` and `Resident` schema is to replace `…Resident…` to `…ResidentUser…`.

File: apps/condo/domains/organization/utils/serverSchema/guards.js [hunk 2]
```diff
@@ -47,7 +47,7 @@ const checkUserExistency = async (context, email, phone) => {
         return usersByEmail[0]
     }
 
-    const usersByPhone = await User.getAll(context, { phone })
+    const usersByPhone = await User.getAll(context, { phone, type: 'staff' })
```
[the remark this answers is not in the document]
Comment (dkoviazin) [reply apps/condo/domains/organization/utils/serverSchema/guards.js:50]: Done

File: apps/condo/domains/ticket/access/TicketFile.js [hunk 3]
```diff
@@ -5,13 +5,18 @@
...
+
+async function canReadTicketFiles ({ authentication: { item: user }, originalInput }) {
     if (!user) return false
     if (user.isAdmin) {
         return {}
     }
+    if (user.type === 'resident') {
+        if (user.id === get(originalInput, ['createdBy', 'id'])) {
+            return true
+        }
```
[the remark this answers is not in the document]
Comment (dkoviazin) [reply apps/condo/domains/ticket/access/TicketFile.js:18]: But we already have there:
```js
return { organization: { employees_some: { user: { id: user.id }, isBlocked: false } } }
```

File: apps/condo/domains/user/schema/index.js [hunk 4]
```diff
@@ -19,5 +21,7 @@ module.exports = {
     ForgotPasswordService,
     ConfirmPhoneAction,
     ConfirmPhoneActionService,
+    SigninOrRegisterResidentService,
+    ChangePhoneNumberResidentService,
```
[the remark this answers is not in the document]
Comment (dkoviazin) [reply apps/condo/domains/user/schema/index.js:25]: Done

File: apps/condo/domains/user/schema/ChangePhoneNumberResidentUserService.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+                        token,
+                        completedAt: null,
+                        isPhoneVerified: true,
+                    }
+                )
+                if (!action) {
+                    throw new Error(`${CONFIRM_PHONE_ACTION_EXPIRED}] Unable to find confirm phone action`)
+                }
+                const { phone: newPhone } = action
+                await User.update(context, userId, { phone: newPhone })
```
Comment (Dimitreee) [inline apps/condo/domains/user/schema/ChangePhoneNumberResidentUserService.js:43]: add phone normalization/validation

File: apps/condo/domains/user/schema/RegisterNewUserService.test.js [hunk 6]
```diff
@@ -1,9 +1,37 @@
 const faker = require('faker')
-const { createTestUser, registerNewUser, createTestPhone, createTestEmail } = require('@condo/domains/user/utils/testSchema')
+const { createTestUser, registerNewUser, createTestPhone, createTestEmail, createTestConfirmPhoneAction } = require('@condo/domains/user/utils/testSchema')
 const { REGISTER_NEW_USER_MUTATION } = require('@condo/domains/user/gql')
 const { EMAIL_ALREADY_REGISTERED_ERROR, PHONE_ALREADY_REGISTERED_ERROR } = require('@condo/domains/user/constants/errors')
 const { makeLoggedInAdminClient, makeClient } = require('@core/keystone/test.utils')
 
+
+describe('RegisterNewUserService createdAt bug', () => {
```
Comment (Dimitreee) [inline apps/condo/domains/user/schema/RegisterNewUserService.test.js:8]: please add more readable test suite name

File: apps/condo/domains/user/schema/SigninOrRegisterResidentUserService.js [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+        {
+            access: true,
+            type: 'type SigninOrRegisterResidentUserOutput { item: User, token: String! }',
+        },
+    ],
+
+    mutations: [
+        {
+            access: access.canSigninOrRegisterResidentUser,
+            schema: 'signinOrRegisterResidentUser(data: SigninOrRegisterResidentUserInput!): SigninOrRegisterResidentUserOutput',
```
Comment (Dimitreee) [inline apps/condo/domains/user/schema/SigninOrRegisterResidentUserService.js:27]: why should we use 1 mutation for signin and register? i'm not sure this is good idea

File: apps/condo/domains/user/schema/ChangePhoneNumberResidentUserService.js (hunk 5 above)
Comment (dkoviazin) [reply apps/condo/domains/user/schema/ChangePhoneNumberResidentUserService.js:44]: This phone is from startConfirmPhoneAction mutation it has normalize there - here we only fetch it from database

File: apps/condo/domains/user/schema/SigninOrRegisterResidentUserService.js (hunk 7 above)
Comment (dkoviazin) [reply apps/condo/domains/user/schema/SigninOrRegisterResidentUserService.js:26]: This is to satisfy mobile app developers. They only have 1 screen to login and register

**Review by AntonAL — APPROVED** [review]

File: apps/condo/domains/user/schema/RegisterNewUserService.test.js (hunk 6 above)
Comment (dkoviazin) [reply apps/condo/domains/user/schema/RegisterNewUserService.test.js:8]: It was a temporary test to catch a bug. 
Removed

**Review by sitozzz — APPROVED** [review]
Looks great 👍

File: apps/condo/domains/ticket/access/TicketFile.js [hunk 8]
```diff
@@ -5,13 +5,18 @@
...
- 
 
-async function canReadTicketFiles ({ authentication: { item: user } }) {
+
+async function canReadTicketFiles ({ authentication: { item: user }, originalInput }) {
     if (!user) return false
     if (user.isAdmin) {
         return {}
     }
+    if (user.type === 'resident') {
```
[the remark this answers is not in the document]
Comment (dkoviazin) [reply apps/condo/domains/ticket/access/TicketFile.js:15]: To allow resident user create ticket file or ticket we need first enable connection from resident user to organization

File: apps/condo/domains/ticket/access/TicketFile.js [hunk 9]
```diff
@@ -20,6 +25,10 @@ async function canManageTicketFiles ({ authentication: { item: user }, originalI
     if (user.isAdmin) return true
     if (operation === 'create') {
         const organizationIdFromTicketFile = get(originalInput, ['organization', 'connect', 'id'])
+        // TODO(zuch): need to check connection from resident to organization
+        if (user.type === 'resident' ) {
```
[the remark this answers is not in the document]
Comment (dkoviazin) [reply apps/condo/domains/ticket/access/TicketFile.js]: To allow resident user create ticket file or ticket we need first enable connection from resident user to organization

File: apps/condo/domains/user/schema/SigninOrRegisterResidentUserService.js (hunk 7 above)
Comment (Dimitreee) [reply apps/condo/domains/user/schema/SigninOrRegisterResidentUserService.js:28]: Not sure if the UI should affect the API, it's better to separate the handles.

**Review by Dimitreee — APPROVED** [review]

File: apps/condo/domains/user/schema/RegisterNewUserService.js [hunk 10]
```diff
@@ -49,34 +37,40 @@ const RegisterNewUserService = new GQLCustomSchema('RegisterNewUserService', {
                 const { confirmPhoneActionToken, ...restUserData } = data
                 const userData = {
                     ...restUserData,
+                    type: 'staff',
```
Comment (pahaz) [inline apps/condo/domains/user/schema/RegisterNewUserService.js:40]: it's better to create and use USER_TYPE const

File: apps/condo/domains/user/schema/RegisterNewUserService.js [hunk 11]
```diff
@@ -5,31 +5,19 @@ const { COUNTRIES } = require('@condo/domains/common/constants/countries')
...
-        variables: { where: { [field]: value } },
-    })
-
-    if (errors) {
-        throw new Error(`[error] Unable to check field ${field} uniques`)
-    }
-
-    if (data.objs.length !== 0) {
-        throw new Error(`[unique:${field}:multipleFound] ${models} with this ${field} is already exists`)
+    const existed = await UserServerUtils.getAll(context, { [field]: value, type: 'staff' })
```
Comment (pahaz) [inline apps/condo/domains/user/schema/RegisterNewUserService.js:18]: it's better to create and use USER_TYPE const

File: apps/condo/domains/user/schema/RegisterNewUserService.js [hunk 12]
```diff
@@ -5,31 +5,19 @@ const { COUNTRIES } = require('@condo/domains/common/constants/countries')
...
 const { MIN_PASSWORD_LENGTH } = require('@condo/domains/user/constants/common')
-const { ConfirmPhoneAction } = require('@condo/domains/user/utils/serverSchema')
+const { ConfirmPhoneAction: ConfirmPhoneActionServerUtils, User: UserServerUtils } = require('@condo/domains/user/utils/serverSchema')
 const isEmpty = require('lodash/isEmpty')
+const {
+    CONFIRM_PHONE_ACTION_EXPIRED,
+} = require('@condo/domains/user/constants/errors')
 
-async function ensureNotExists (context, model, models, field, value) {
+async function ensureNotExists (context, field, value) {
```
Comment (pahaz) [inline apps/condo/domains/user/schema/RegisterNewUserService.js:14]: ensureStaffUserNotExists

File: apps/condo/domains/user/schema/RegisterNewUserService.js [hunk 13]
```diff
@@ -49,34 +37,40 @@ const RegisterNewUserService = new GQLCustomSchema('RegisterNewUserService', {
...
+                await ensureNotExists(context, 'phone', userData.phone)
+                await ensureNotExists(context, 'email', userData.email)
 
                 if (userData.password.length < MIN_PASSWORD_LENGTH) {
                     throw new Error(`${MIN_PASSWORD_LENGTH_ERROR}] Password length less then ${MIN_PASSWORD_LENGTH} character`)
                 }
-                const { data: createData, errors: createErrors } = await context.executeGraphQL({
+                // TODO(zuch): fix bug when user can not be created becaues of createAt and updatedAt fields
+                // const user = await UserServerUtils.create(context, userData)
+                const { data: { result: user }, errors: createErrors } = await context.executeGraphQL({
```
Comment (pahaz) [inline apps/condo/domains/user/schema/RegisterNewUserService.js:69]: why not just User.create( ... ) ? (optional)

File: apps/condo/domains/user/schema/SigninResidentUserService.js [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+                    }
+                )
+                if (!action) {
+                    throw new Error(`${CONFIRM_PHONE_ACTION_EXPIRED}] Unable to find confirm phone action`)
+                }
+                userData.phone = action.phone
+                userData.type = 'resident'
+                userData.isPhoneVerified = true
+                userData.password = getRandomString()
+                let [user] = await UserServerUtils.getAll(context, { type: 'resident', phone: userData.phone })
```
Comment (pahaz) [inline apps/condo/domains/user/schema/SigninResidentUserService.js:52]: we should throw an exception if we get more than one object

**Pull request merged**